### ISSUE TITLE
Different holder type unchecked in argument loading

### ISSUE THREAD

**Issue by YannickJadoul** [pybind/pybind11#751]
Though I know it might be considered a case of "stupid code, don't write that", I just found the following way of segfaulting a pybind11 extension module:

```C++
#include <pybind11/pybind11.h>
#include <iostream>

struct Test {
	int x = 42;
};

namespace py = pybind11;

PYBIND11_PLUGIN(test) {
	py::module m("test");
	
	py::class_<Test /* implicitly held by std::unique_ptr */>(m, "Test")
		.def(py::init<>())
	;
	
	m.def("f", [] (std::shared_ptr<Test> test) {
		std::cout << "Here" << std::endl;
		std::cout << test->x << std::endl;
		std::cout << "There" << std::endl;
	});
	
	return m.ptr();
}
```

Calling function `f` with an argument of type `Test` does not result in a type exception, but segfaults Python:

```Python
>>> import test
>>> test.f(test.Test())
Here
42
There
Segmentation fault (core dumped)
```
(Note that you might need to repeat this, as it seems that sometimes the 'undefined behaviour' causes nothing to happen. I am guess this is the case because the layouts of `std::unique_ptr` and `std::shared_ptr` overlap and so the pointer itself is not invalid.)

As far as I see, the problem lies in the fact that 'copyable_holder_caster' does not check (and has no way of checking?) the type of the holder once the held type matches (i.e., [this line in cast.h](https://github.com/pybind/pybind11/blob/0f5ec0a87e3d3879cd70740422adf3e3728a9d84/include/pybind11/cast.h#L959). This is actually how I found this way of segfaulting, by trying to understand the casting mechanism, rather than writing such stupid code.

I am aware you have to make a rather severe programming mistake to trigger this, but it does not seem not to match the philosophy of pybind11 regarding type safety, and since it could not find a previous bug report on this, I reckoned this might be interesting.

**Comment by jagerman** [comment]
I haven't investigated yet, but just from reading your report it sounds like a legitimate bug to me.

**Comment by dean0x7d** [comment]
Indeed, it's an issue. Luckily, #588 added the information needed to check for this (at least for the basic default/non-default holder case). I'll prepare a fix.

**Comment by YannickJadoul** [comment]
Ok, thanks!

Depending on the amount of time you have available, I would also be willing to look into this and submit a pull request, if you prefer.

**Comment by dean0x7d** [comment]
Fixed with #753. Thanks for the report, @YannickJadoul!
